### ISSUE TITLE
Frame latency rework worsens frame time variance on at least Nvidia

### ISSUE THREAD

**Issue by aufkrawall** [doitsujin/dxvk#2281]
With 0fe8f2e40cf4e7a9aec744ce00c502bb1d1a3100 , frame time graph in e.g. Assassin's Creed Odyssey (it seems to affect every game) is mostly flat (ignore the game's own crazy graph):
![ACOdyssey_2021_09_08_20_26_15_861](https://user-images.githubusercontent.com/13510490/132576618-93ae4e06-5f3d-49ab-add2-85a300ed7afc.png)

With 747834e9b02288f9ab7a2514fdffd85ad78db6ab , it is way bumpier and should be noticeably worse (not much, but it seems to sum up with the game's own slight stutter):
![ACOdyssey_2021_09_08_21_45_43_550](https://user-images.githubusercontent.com/13510490/132576858-b1de6f31-0e35-4246-965a-3d21786a217f.png)

It doesn't help to set `dxgi.maxFrameLatency` to different values. E.g. before the rework, `1` produced better frame times than `2` after the rework. With the Nvidia D3D11 driver, ULL set to on (equals `1`) looks exactly as smooth as with DXVK before the rework. So only after the rework frame time variance is worse than with native D3D11 (AMD D3D11 produces only crap results here btw.).

### Software information
I tested Odyssey and Hitman 2. I think you need to be GPU bound, so e.g. use SSAA resolution scale option in Hitman 2. This is all with vsync off.

### System information
- GPU:  RTX 3060
- Driver: 471.96 / 470.62.02
- Wine version: recent staging-tkg / Windows 10
- DXVK version: Issue started with 1.9 accordingly.

### Log files

[hitman2_dxgi.log](https://github.com/doitsujin/dxvk/files/7131524/hitman2_dxgi.log)
[hitman2_d3d11.log](https://github.com/doitsujin/dxvk/files/7131525/hitman2_d3d11.log)

It might be that AMDVLK/RADV didn't show this issue, but I can't really recall.

**Comment by doitsujin** [comment]
This is known, and I don't know how to work around this issue.

**Comment by aufkrawall** [comment]
Thanks. The workaround mentioned in that ticket to revert e64a8ee61a0669b317043bb211f1c519bd4c47af unfortunately doesn't do the trick , as it seems to effectively limit GPU load at ~70% (no CPU prerender at all?). :(

**Comment by doitsujin** [comment]
Does [this](https://github.com/doitsujin/dxvk/tree/nv-test) branch improve things?

**Comment by aufkrawall** [comment]
The issue unfortunately remains unchanged with this branch.

**Comment by doitsujin** [comment]
Not sure what to do then other than asking people who have this issue to stay on 1.8.1 for the time being.

I cannot reliably reproduce this on my Nvidia laptop, so debugging this is pretty much impossible.

**Comment by aufkrawall** [comment]
Probably not the end of the world to endure with 1.8.1 in most cases. :)

Hopefully Nvidia devs can investigate why their driver doesn't produce an optimal result in this case.

**Comment by aufkrawall** [comment]
Unfortunately still unchanged with latest git-master and Nvidia driver 511.79 (includes the Vulkan 1.3 changes):
![ACOdyssey_2022_02_21_19_18_31_835](https://user-images.githubusercontent.com/13510490/155009157-f81b8c70-a460-428b-b732-5966279e698a.png)

Very ugly and unpopular notion: Perhaps it would make sense now to additionally re-implement the old presentation path without the latency optimizations and automatically switch to that when an Nvidia GPU is detected (+config option)? * takes cover *

**Comment by doitsujin** [comment]
I don't want to maintain multiple paths for this, and so far there has not been any evidence that this is a widespread problem even on Nvidia hardware, no one else has observed degraded frame times yet.

I pushed something [here](https://github.com/doitsujin/dxvk/tree/late-acquire) which does something similar to our old path, please check if that helps.

**Comment by aufkrawall** [comment]
I'd expect this to occur on any system with that driver (but not any game), it's just hard to notice without VRR (and lots of games tend to have a bit of microstutter anyway without forcing maxFrameLatency 1). I probably also encountered it in Black Mesa, but it seems we're at a point where more testing doesn't make any more sense and can just call it a day then.

**Comment by K0bin** [comment]
I played a bit of Black Mesa recently and I'm pretty sure I did not have this problem. I even play with the DXVK frame time graph enabled, so I would've definitely noticed.

**Comment by aufkrawall** [comment]
I just have to load the game's menu to reproduce (1.10 and 1.8.1, not caused by background shader compiling etc., also `d3d9.maxFrameLatency` doesn't matter):
![bms_2022_03_04_20_11_10_165](https://user-images.githubusercontent.com/13510490/156827033-ea27fc52-6f03-460e-a26f-0eb9abc844ec.png)
![bms_2022_03_04_20_12_09_837](https://user-images.githubusercontent.com/13510490/156827037-38d77364-756b-474f-8cb5-4c063dd30a16.png)

Both DXVK's and MangoHud's frame time graph by default are a bit too insensitive, so it's easy to miss (it's RTSS own graph in the screenshots, which by default is also more sensitive than Afterburner's). And as I said earlier, the generally higher variance is not that problematic for a lot of the time, but it can sum up enough in some situations to be noticeable. It's just harder to reproduce and to capture it isolated in a screenshot. Also no difference Linux vs. Windows. Really doubt some Nvidia users would be affected while others are not when I can bisect it to those specific commits.

Guess I'll just stick with 1.8.1 when it's a "won't fix".

**Comment by doitsujin** [comment]
I just don't see a way to figure out the root cause without anyone being able to reproduce it, and I don't see anything wrong with our code either.

**Comment by doitsujin** [comment]
FWIW, Black Mesa on Windows + DXVK looks like this here:
![Black Mesa_2022 03 05-17 52](https://user-images.githubusercontent.com/25567304/156892617-42f8d871-8355-472e-be4b-5b9d2be1c8b4.png)

**Comment by aufkrawall** [comment]
Thanks for checking!
Though it seems you're CPU bound in that scene, at least according to the GPU usage stats. When I decrease resolution to get CPU bound, there also is no difference for me in frame time variance with 1.8.1 vs. 1.10:
![bms_2022_03_05_18_06_29_298](https://user-images.githubusercontent.com/13510490/156893225-3f6e6e64-131b-43e9-86e1-4c10a1812541.png)

![bms_2022_03_05_18_07_40_280](https://user-images.githubusercontent.com/13510490/156893226-ddabcb86-fea3-4c08-b4d3-d45442ef5fb8.png)

When GPU bound for the issue to occur, it's also visible in DXVK's graph, though it looks far less dramatic:
![362890_20220305181413_1](https://user-images.githubusercontent.com/13510490/156893411-a86a8d05-102a-4e2f-aa76-d2d11f4568f6.png)

![362890_20220305181533_1](https://user-images.githubusercontent.com/13510490/156893415-ceb7ae86-83e8-4cd0-935f-b7d7007bce3a.png)

So it seems it's not like DXVK's stats wouldn't catch it, it's just the scale of the graph.

Btw. If you try to get GPU bound by forcing higher resolution via VSR on Windows: That might not work out too well, even with native D3D9 the 32 bit game doesn't seem to be able to run in 5k resolution without running out of address space (main menu might still work though).

I can't judge how much effort it would require, but would it be possible to provide an unofficial patch that reverts the latency optimizations for the time being? I tried reverting the commits myself, but that didn't work out.

**Comment by doitsujin** [comment]
That's exactly why I don't want to do it, it's a pain in the arse and overall the patches seem beneficial (and are required for Gamescope's frame rate limiter to work properly without introducing ludicrous amounts of latency).

**Comment by aufkrawall** [comment]
Looks like something's mitigated it, DXVK's graph is a bit smoother now:
![362890_20220821184910_1](https://user-images.githubusercontent.com/13510490/185802085-be0e6fcf-1d75-4133-8a71-2c40dde1eba4.png)

I think we can close this and if someone's found a game where this is more dramatic, open a new issue for it.
It's also interesting that on Windows with RTSS, changing its statistics setting to either frame start or frame presentation can give vastly different results, especially with native D3D11 and Nvidia driver ultra low latency ultra setting. So I guess if we wanted to be super exact, we would need to apply FCAT after display scan-out with a dedicated capture system in between, as apparently software overlays aren't 100% trustworthy.
